lib/fizzy/parser_expr.cpp

```diff
@@ -340,10 +381,28 @@ parser_result<Code> parse_expr(const uint8_t* pos, const uint8_t* end, FuncIdx f
             break;
 
         case Instr::drop:
+            drop_operand(frame, operand_stack, OperandStackType::Unknown);
+            break;
         case Instr::select:
-            // TODO: for select validate that two top operands are the same type
-            drop_operand(frame, operand_stack, std::nullopt);
+        {
+            const auto frame_stack_height = static_cast<int>(operand_stack.size());
+
+            // Two operands are expected, because the selector operand was already popped
+            // according to instruction type table
+            if (!frame.unreachable && frame_stack_height < frame.parent_stack_height + 2)
+                throw validation_error{"stack underflow"};
+
+
+            const auto operand_type = frame_stack_height > frame.parent_stack_height ?
+                                          operand_stack[0] :
+                                          OperandStackType::Unknown;
```

Comment:
So we get unknown in case of unreachable?

Reply:
Yes, if unreachable and nothing is pushed before, then `select` doesn't know types and will push `Unknown` in the end.

Reply:
This comes from "polymorphic stack" concept in wasm. Even if unreachable, some information about instructions is validated.